### PR TITLE
ImageObjectFunctorial/CoimageObjectFunctorial

### PR DESCRIPTION
**Pull request by mohamed-barakat**

### CODE REVIEW

**Comment by zickgraf** [comment]
Some thoughts: CAP seems to be inconsistent with its naming convention: `ImageObject` vs. `Coimage`. Maybe because `Image` is already occupied by GAP? But then again, `Cokernel` is not occupied by GAP but the object is still called `CokernelObject`. In any case, I think the names `ImageObjectFunctorial` and `KernelObjectFunctorial` are a bit weird, because the whole construction is functorial, not the object itself. Should we call it `ImageFunctorial`, and maybe at some point rename `KernelObjectFunctorial` to `KernelFunctorial`?

Ping @sebastianpos

**Comment by sebastianpos** [comment]
Yes, it should be `CoimageObject` instead of `Coimage`. This is an old inconsistency.

I can explain my reasoning when it comes to the naming `KernelObjectFunctorial` (instead of `KernelFunctorial`):
In a truly categorical programming language, all data types X should be (1)-categories, and all functions f:X->Y between data types should be functors. In particular, f should be applicable both to objects and morphisms of the data type (=category) X.

Since GAP is not a categorical programming language, we need an extra function for the action of f on morphisms. My thought was that such functions should be called `fFunctorial`.

In particular, `KernelObject`, `KernelEmbedding`, `KernelLift` could all be extended to functors, i.e., not only `KernelObjectFunctorial` makes sense, but also `KernelEmbeddingFunctorial` and `KernelLiftFunctorial`.

Since this idea did not turn out as important as I thought in the beginning, I'm open to a change.

**Comment by mohamed-barakat** [comment]
Maybe Sebastian's overall functoriality might become relevant in the future. For this, I would prefer `CoimageObjectFunctorial`. Shall I wait for `CAP` to change `Coimage` -> `CoimageObject`? Otherwise, I can prepare for this switch and do the renaming here accordingly. In any case, I want this code to be part of `CAP` in the future (This package is my playground for features I need without delay).

**Comment by zickgraf** [comment]
Thanks for the thorough explanation @sebastianpos! With this I'm happy with the current names.

@mohamed-barakat I don't see a downside of naming the functorial `CoimageObjectFunctorial`. From grepping through packages it looks like there aren't many occurrences of `Coimage`, so maybe I do the renaming in the near future.

File: gap/Tools.gd [hunk 1]
```diff
@@ -14,6 +14,67 @@ DeclareGlobalFunction( "RELATIVE_WEAK_BI_FIBER_PRODUCT_PREFUNCTION" );
...
+##
+###################################
+
+#! @Description
+#! The arguments are three morphisms
+#! $\alpha: A \rightarrow B, \mu: A \rightarrow A', \alpha': A' \rightarrow B'$.
+#! The output is the morphism
+#! $\mathrm{Coimage}( \alpha ) \rightarrow \mathrm{Coimage}( \alpha' )$
+#! given by the functoriality of the coimage.
+#! @Returns a morphism in $\mathrm{Hom}(\mathrm{Coimage}( \alpha ), \mathrm{Coimage}( \alpha' ))$
```
Comment (zickgraf) [inline gap/Tools.gd:59]:
```suggestion
#! @Returns a morphism in $\mathrm{Hom}(\mathrm{CoimageObject}( \alpha ), \mathrm{CoimageObject}( \alpha' ))$
```

File: gap/Tools.gd [hunk 2]
```diff
@@ -14,6 +14,67 @@ DeclareGlobalFunction( "RELATIVE_WEAK_BI_FIBER_PRODUCT_PREFUNCTION" );
...
+#! @Arguments alpha, mu, alpha_prime
+DeclareOperation( "CoimageObjectFunctorial",
+        [ IsCapCategoryMorphism, IsCapCategoryMorphism, IsCapCategoryMorphism ] );
+
+
+#! @Description
+#! The arguments are an object $s = \mathrm{Coimage}( \alpha )$,
+#! three morphisms
+#! $\alpha: A \rightarrow B, \mu: A \rightarrow A', \alpha': A' \rightarrow B'$,
+#! and an object $r = \mathrm{Coimage}( \alpha' )$.
```
Comment (zickgraf) [inline gap/Tools.gd:69]:
```suggestion
#! and an object $r = \mathrm{CoimageObject}( \alpha' )$.
```

File: gap/Tools.gd [hunk 3]
```diff
@@ -14,6 +14,67 @@ DeclareGlobalFunction( "RELATIVE_WEAK_BI_FIBER_PRODUCT_PREFUNCTION" );
...
+#! The output is the morphism
+#! $\mathrm{Coimage}( \alpha ) \rightarrow \mathrm{Coimage}( \alpha' )$
```
Comment (zickgraf) [inline gap/Tools.gd:71]:
```suggestion
#! $\mathrm{Coimage}( \alpha ) \rightarrow \mathrm{CoimageObject}( \alpha' )$
```

**Review by zickgraf — CHANGES_REQUESTED** [review]

**Pull request merged**